File: backend-java/src/main/java/com/feedefaulter/services/OcrService.java
package com.feedefaulter.services;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.cdimascio.dotenv.Dotenv;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.File;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class OcrService {

    private final String apiKey;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public OcrService() {
        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = new ObjectMapper();

        // Load dotenv from root project dir
        String key = null;
        try {
            Dotenv dotenv = Dotenv.configure()
                    .directory("../")
                    .ignoreIfMalformed()
                    .ignoreIfMissing()
                    .load();
            key = dotenv.get("GEMINI_API_KEY");
        } catch (Exception e) {
            System.out.println("[WARN] Dotenv load failed, trying system env: " + e.getMessage());
        }

        if (key == null) {
            key = System.getenv("GEMINI_API_KEY");
        }

        this.apiKey = key;
        System.out.println("[INIT] OcrService loaded API key: " + (apiKey != null ? "FOUND" : "NOT FOUND"));
    }

    public Map<String, Object> extractReceiptData(MultipartFile file) {
        try {
            byte[] fileBytes = file.getBytes();
            String mimeType = file.getContentType();
            if (mimeType == null) {
                mimeType = "image/jpeg";
            }
            return extractReceiptDataFromBytes(fileBytes, mimeType);
        } catch (Exception e) {
            System.err.println("[ERROR] Failed to read multipart file: " + e.getMessage());
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to read file: " + e.getMessage());
            return error;
        }
    }

    public Map<String, Object> extractReceiptDataFromPath(String filePath) {
        try {
            File localFile = new File(filePath);
            byte[] fileBytes = Files.readAllBytes(localFile.toPath());
            String mimeType = Files.probeContentType(localFile.toPath());
            if (mimeType == null) {
                mimeType = "image/jpeg";
            }
            return extractReceiptDataFromBytes(fileBytes, mimeType);
        } catch (Exception e) {
            System.err.println("[ERROR] Failed to read local file: " + e.getMessage());
            Map<String, Object> error = new HashMap<>();
            error.put("error", "Failed to read file: " + e.getMessage());
            return error;
        }
    }

    private Map<String, Object> extractReceiptDataFromBytes(byte[] fileBytes, String mimeType) {
        Map<String, Object> result = new HashMap<>();

        if (apiKey == null || apiKey.isEmpty()) {
            System.out.println("[WARN] No Gemini API key found, returning mock OCR data");
            result.put("transaction_id", "MOCK_UTR_" + System.currentTimeMillis());
            result.put("amount", 5000);
            result.put("date", "31/05/2026");
            result.put("confidence", "High");
            return result;
        }

        try {
            String base64Data = Base64.getEncoder().encodeToString(fileBytes);

            String prompt = "Analyze this bank transaction receipt/challan.\n" +
                    "Extract the following information and return ONLY a valid JSON object. Do not include markdown formatting or extra text.\n\n" +
                    "Required JSON format:\n" +
                    "{\n" +
                    "    \"transaction_id\": \"Extract the 12 to 16 digit UTR or Transaction/Reference Number\",\n" +
                    "    \"amount\": \"Extract the numeric amount paid (just the number without commas or currency symbols)\",\n" +
                    "    \"date\": \"Extract the date of transaction in DD/MM/YYYY format\",\n" +
                    "    \"confidence\": \"High, Medium, or Low depending on how clearly you can read it\"\n" +
                    "}\n\n" +
                    "If a field is missing or unreadable, set its value to null.";

            // Build request body manually or using map
            Map<String, Object> inlineData = Map.of(
                    "mime_type", mimeType,
                    "data", base64Data
            );
            Map<String, Object> part1 = Map.of("text", prompt);
            Map<String, Object> part2 = Map.of("inline_data", inlineData);
            Map<String, Object> content = Map.of("parts", List.of(part1, part2));
            Map<String, Object> requestBody = Map.of("contents", List.of(content));

            String jsonPayload = objectMapper.writeValueAsString(requestBody);

            String url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=" + apiKey;

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(jsonPayload))
                    .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new RuntimeException("API request failed with status: " + response.statusCode() + " " + response.body());
            }

            Map<?, ?> jsonResponse = objectMapper.readValue(response.body(), Map.class);
            List<?> candidates = (List<?>) jsonResponse.get("candidates");
            if (candidates == null || candidates.isEmpty()) {
                throw new RuntimeException("No candidates returned from Gemini");
            }

            Map<?, ?> candidate = (Map<?, ?>) candidates.get(0);
            Map<?, ?> responseContent = (Map<?, ?>) candidate.get("content");
            List<?> parts = (List<?>) responseContent.get("parts");
            if (parts == null || parts.isEmpty()) {
                throw new RuntimeException("No parts returned from Gemini candidate");
            }

            Map<?, ?> part = (Map<?, ?>) parts.get(0);
            String rawText = ((String) part.get("text")).strip();

            // Clean markdown block wrappers if present
            if (rawText.startsWith("```json")) {
                rawText = rawText.replace("```json", "");
            } else if (rawText.startsWith("```")) {
                rawText = rawText.replace("```", "");
            }
            if (rawText.endsWith("```")) {
                rawText = rawText.substring(0, rawText.length() - 3);
            }

            rawText = rawText.strip();
            Map<?, ?> ocrData = objectMapper.readValue(rawText, Map.class);

            // Populate output result map
            result.put("transaction_id", ocrData.get("transaction_id"));
            result.put("amount", ocrData.get("amount"));
            result.put("date", ocrData.get("date"));
            result.put("confidence", ocrData.get("confidence"));

            return result;

        } catch (Exception e) {
            System.err.println("[ERROR] OCR Processing failed: " + e.getMessage());
            result.put("error", e.getMessage());
            // Fallback mock values
            result.put("transaction_id", "MOCK_UTR_" + System.currentTimeMillis());
            result.put("amount", 5000);
            result.put("date", "31/05/2026");
            result.put("confidence", "Low");
            return result;
        }
    }
}

File: backend-python/services/ocr_service.py
import os
import json
import time
from google import genai
from google.genai import types

# Models to try in order (primary → fallback)
GEMINI_MODELS = ['gemini-2.5-flash', 'gemini-1.5-flash', 'gemini-1.5-flash-8b']
MAX_RETRIES = 3
RETRY_DELAY = 2  # seconds between retries

def _parse_json_response(raw_text):
    """Parse JSON from Gemini response, handling markdown wrappers."""
    raw_text = raw_text.strip()
    if raw_text.startswith("```json"):
        raw_text = raw_text.replace("```json", "", 1)
    if raw_text.startswith("```"):
        raw_text = raw_text.replace("```", "", 1)
    if raw_text.endswith("```"):
        raw_text = raw_text[:-3]
    return json.loads(raw_text.strip())

def extract_receipt_data(image_path):
    """
    Uses Gemini to scan an offline bank receipt and extract structured data.
    Automatically retries with exponential backoff and falls back to a
    lighter model if the primary model is overloaded (503 UNAVAILABLE).
    Returns a dictionary with transaction_id, amount, date, and confidence.
    """
    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        return {"error": "API Key not found."}

    prompt = """
    Analyze this bank transaction receipt/challan.
    Extract the following information and return ONLY a valid JSON object. Do not include markdown formatting or extra text.
    
    Required JSON format:
    {
        "transaction_id": "Extract the 12 to 16 digit UTR or Transaction/Reference Number",
        "amount": "Extract the numeric amount paid (just the number without commas or currency symbols)",
        "date": "Extract the date of transaction in DD/MM/YYYY format",
        "confidence": "High, Medium, or Low depending on how clearly you can read it"
    }
    
    If a field is missing or unreadable, set its value to null.
    """

    client = genai.Client(api_key=api_key)
    last_error = None

    for model_name in GEMINI_MODELS:
        for attempt in range(1, MAX_RETRIES + 1):
            try:
                print(f"[OCR] Trying model={model_name}, attempt={attempt}/{MAX_RETRIES}")

                # Upload image to Gemini File API
                sample_file = client.files.upload(file=image_path)

                response = client.models.generate_content(
                    model=model_name,
                    contents=[sample_file, prompt]
                )

                # Clean up uploaded file from Google's servers
                try:
                    client.files.delete(name=sample_file.name)
                except Exception:
                    pass

                data = _parse_json_response(response.text)
                print(f"[OCR] Success with model={model_name} on attempt={attempt}")
                return data

            except Exception as e:
                last_error = e
                error_str = str(e)
                print(f"[OCR] Error (model={model_name}, attempt={attempt}): {error_str}")

                # 503 / UNAVAILABLE → retry with backoff, then try next model
                if "503" in error_str or "UNAVAILABLE" in error_str or "high demand" in error_str.lower():
                    if attempt < MAX_RETRIES:
                        wait = RETRY_DELAY * attempt  # 2s, 4s, 6s ...
                        print(f"[OCR] Model busy. Waiting {wait}s before retry...")
                        time.sleep(wait)
                        continue
                    else:
                        print(f"[OCR] All {MAX_RETRIES} retries exhausted for {model_name}. Trying next model...")
                        break  # Move to next model

                # JSON parse error → no point retrying, return gracefully
                elif isinstance(e, (json.JSONDecodeError, ValueError)):
                    print("[OCR] JSON parse error. Returning raw text.")
                    return {"error": "Could not parse AI response. Please try again."}

                # Other errors → break out and try next model
                else:
                    break

    print(f"[OCR] All models failed. Last error: {last_error}")
    return {"error": "AI scan failed due to high demand. Please try again in a moment."}
